Fix:- Improve HOC support and state preservation in React Refresh

Biki-das opened the pull request:
## Summary

This fixes #30659 , the issue was how the state was preserved and needed special cases for the forward and memo, have also added tests related to the same.
## How did you test this change?

`yarn test packages/react-refresh/src/__tests__/ReactFresh-test.js`

![Screenshot 2024-08-12 at 4 27 39 PM](https://github.com/user-attachments/assets/2b597a62-c45f-443b-acfc-3232962ba0a3)

Biki-das commented:
cc @eps1lon @hoxyq 
would love to know thoughts from all of you!

Biki-das commented:
@gaearon did you encounter this error? if yes could you help me out if this fix looks right.

Biki-das commented:
cc @josephsavona 
Might be something to fix. would love to know your thoughts

hoxyq commented:
Hey @Biki-das, apologies for a long wait. I think we should find an owner for `react-refresh`, I am not sure when was the last time `react-refresh` was actually updated.

Biki-das commented:
@hoxyq any update on this, 
may be @poteto

hoxyq commented:
> @hoxyq any update on this, 
> 
> may be @poteto 

Hey, @Biki-das, no updates from me, I am currently on leave and will be back in December.

Biki-das commented:
@hoxyq whenever you return and if you are a bit less busy, could you forward this further to someone who might have idea about react refresh, feel like this fix could be important.

Biki-das commented:
@eps1lon any thoughts on who can review this, i see react refresh has not been touched for a long time? could we get someone to review this if you know someone.

hoxyq reviewed (commented):
Overall looks good, and changes make sense, few points:
- Could you please retitle this PR to something that describes what's being added to Fresh?
- Could you please add tests that will check that state is not preserved when switching between different wrappers?

Biki-das commented:
@hoxyq thanks for the feedback sure i will try to add the required tests and change the PR title as well.

Biki-das commented:
@hoxyq let me know if the new tests added are what you expected, and the reset test is running fine, we have a playwright test fail which is non related to this commit. looking forward to your thoughts

Biki-das commented:
> Overall looks good, and changes make sense, few points:
> 
> * Could you please retitle this PR to something that describes what's being added to Fresh?
> * Could you please add tests that will check that state is not preserved when switching between different wrappers?

also nothing new is added to Fresh, these are the changes

Added special handling for React.forwardRef and React.memo components
Enhanced state preservation logic between updates
Added checks for component type signatures and React component types
Improved family registration for HOCs

Biki-das commented:
@hoxyq friendly ping.

Biki-das reviewed (commented):
Up for final review

hoxyq commented:
Thanks!

gaearon commented:
Why is this necessary? What set of tooling does that reproduce with?

These are supposed to be handled by the module bundler. If we switch between a function and memo, I would expect “is exporting only functions” to switch from true to false. From what I remember, in that case the modules that depend on this code are expected to start using the new versions directly (without trying to proxy the implementations). So it should “just work”.

I’m not convinced this is a correct or necessary fix. I would need to read more closely through the code to say for sure. But I would encourage to check the React Native (Metro) or the webpack plugin behavior instead of testing this in isolation. If they handle it with the existing code (which I assume they do — I tested these cases when working on FR), I don’t think this change is needed.

gaearon commented:
If CodeSandbox or Next.js aren’t handling this correctly, I would first assume that this is a problem with _their_ integrations of FR. We need to verify whether this works in the original Metro one.

gaearon commented:
Sorry, i should’ve explicitly marked this as “needs revision” instead of stalling on the review. I sent a revert in https://github.com/facebook/react/pull/32214. We don’t have to merge it right now but I think this needs a more careful analysis.

Biki-das commented:
@gaearon i happened to discover this even using vite with React, so i assumed react refresh to be the issue and just played with the same, tried to see why it is breaking, apologize if this wasn't the correct way since i had limited understanding.

hoxyq commented:
> We need to verify whether this works in the original Metro one.

Switching between `forwardRef` and `memo` wrappers works for me with Metro runtime. Thanks for the explanation.

I am not sure I am following how is the API surface of integration with FR looks like in this case:
> I would expect “is exporting only functions” to switch from true to false. From what I remember, in that case the modules that depend on this code are expected to start using the new versions directly (without trying to proxy the implementations)

I can look up a bit later how Metro does that and check if there is anything special.

Biki-das commented:
thanks for the detailed explanation i tried using CRA as it uses webpack and it seems to have the same issue reproduced. and as i understood what dan meant is module bundlers are expected to handle the case which i introduced, rather than the runtime itself.


https://github.com/user-attachments/assets/6bb4a7df-0157-4464-b135-3f52ea196e66

hoxyq commented:
> @hoxyq did you did the way i did?

Actually, no. With the same code that is in your demo, I can also reproduce this issue. Both with default and named exports.

Previously I would just wrap `Component` in `React.memo` or `React.forwardRef` and export it, but looks like what you actually need is to preserve a name of what you are exporting, and switch it from plain function to wrappers with `memo` or `forwardRef`, like what you are doing in your demo with `Component`.

Biki-das commented:
Yeah that's the exact issue, while editing code, the changes should reflect without any issue!

Biki-das commented:
Have verified the changes that i made in the refresh file whether they actually work or not, as test in isolation might not work, it does work, for CRA i tried configuring it with **crago** and pulled up my `react-refresh-runtime.development.js` file which has the fixes to be used. and there were no errors as we are generally facing.

i have wrapped up a repo to try and test the same, feel free to checkout and let me know if i did something wrong and this is not the way to test the same. 

https://github.com/Biki-das/React-refresh-fix-check



https://github.com/user-attachments/assets/8dcca8d2-946d-40f5-baf1-44e2b93ce854

levrik commented:
Was it intended that this change ended up in today's `0.16.0` release although it was merged after the dev release version was bumped to `0.17.0` already?
